core/src/toga/widgets/slider.py

```diff
@@ -194,48 +221,156 @@ def on_change(self, handler):
 
-        **DEPRECATED: renamed as on_change**
+class SliderImpl(ABC):
+    @abstractmethod
+    def get_value(self):
+        ...
 
-        Returns:
-            The ``callable`` that is executed on slide.
-        """
-        warnings.warn(
-            "Slider.on_slide has been renamed Slider.on_change", DeprecationWarning
-        )
-        return self._on_change
+    @abstractmethod
+    def set_value(self, value):
+        ...
+
+    @abstractmethod
+    def get_range(self):
+        ...
+
+    @abstractmethod
+    def set_range(self, range):
+        ...
+
+    @abstractmethod
+    def get_tick_count(self):
+        ...
+
+    @abstractmethod
+    def set_tick_count(self, tick_count):
+        ...
+
+
+class IntSliderImpl(SliderImpl):
+    """Base class for implementations which use integer values."""
+
+    # Number of steps to use to approximate a continuous slider.
+    CONTINUOUS_MAX = 10000
+
+    def __init__(self):
+        super().__init__()
+
+        # Dummy values used during initialization.
+        self.value = 0
+        self.discrete = False
+
+    def get_value(self):
+        return self.value
+
+    def set_value(self, value):
+        span = self.max - self.min
+        self.set_int_value(round((value - self.min) / span * self.get_int_max()))
+        self.value = value  # Cache the original value so we can round-trip it.
+
+    def get_range(self):
+        return (self.min, self.max)
+
+    def set_range(self, range):
+        self.min, self.max = range
+
+    def get_tick_count(self):
+        return (self.get_int_max() + 1) if self.discrete else None
+
+    def set_tick_count(self, tick_count):
+        if tick_count is None:
+            self.discrete = False
+            self.set_int_max(self.CONTINUOUS_MAX)
+        else:
+            self.discrete = True
+            self.set_int_max(tick_count - 1)
+        self.set_ticks_visible(self.discrete)
+
+    def on_change(self):
+        span = self.max - self.min
+        self.value = self.min + (self.get_int_value() / self.get_int_max() * span)
+        self.interface.on_change(None)
+
+    @abstractmethod
+    def get_int_value(self):
+        ...
+
+    @abstractmethod
+    def set_int_value(self, value):
+        ...
+
+    @abstractmethod
+    def get_int_max(self):
+        ...
+
+    @abstractmethod
+    def set_int_max(self, max):
+        ...
+
+    @abstractmethod
+    def set_ticks_visible(self, visible):
+        ...
+
+
+class ContinuousSliderImpl(SliderImpl):
+    """Base class for implementations which don't support discrete mode. The slider
+    thumb is allowed to move continuously, but the reported value is rounded to the
+    closest tick, as is the thumb itself after it's released.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+        # Dummy values used during initialization.
+        self.value = 0
+        self.tick_count = None
+
+    def get_value(self):
+        return self.value
```

Comment:
As above, is there any reason to prefer a shadow copy here rather than using `self.interface._round_value(self.get_continuous_value())`?

Reply:
You're right, and when I removed the shadow copy, the ContinuousSliderImpl class was so simple that it wasn't worth keeping anymore. However, the iOS implementation still needs a shadow copy for round-tripping, because it's based on 32-bit floats.